core/src/main/scala/caliban/GraphQL.scala

```diff
@@ -4,111 +4,134 @@ import caliban.Rendering.renderTypes
+   * @param that another GraphQL API object
+   * @return a new GraphQL API
    */
-  def provide(r: R): GraphQL[Any, Q, M, S, E] = wrapExecutionWith(_.provide(r))
+  final def combine[R1 <: R](that: GraphQL[R1]): GraphQL[R1] =
+    new GraphQL[R1] {
+      override val schema: RootSchema[R1]                  = self.schema |+| that.schema
+      override val queryAnalyzers: List[QueryAnalyzer[R1]] = self.queryAnalyzers ++ that.queryAnalyzers
+    }
 
   /**
-   * Wraps the `execute` method of the interpreter with the given function.
-   * This can be used to customize errors, add global timeouts or logging functions.
-   * @param f a function from `URIO[R, GraphQLResponse[E]]` to `URIO[R2, GraphQLResponse[E2]]`
-   * @return a new GraphQL interpreter
+   * Operator alias for `combine`.
    */
-  def wrapExecutionWith[R2, E2](
-    f: URIO[R, GraphQLResponse[E]] => URIO[R2, GraphQLResponse[E2]]
-  ): GraphQL[R2, Q, M, S, E2] =
-    new GraphQL[R2, Q, M, S, E2] {
-      override def check(query: String): IO[CalibanError, Unit] = self.check(query)
-      override def execute[R1 <: R2](
-        query: String,
-        operationName: Option[String],
-        variables: Map[String, InputValue],
-        skipValidation: Boolean,
-        queryAnalyzers: List[QueryAnalyzer[R1]] = Nil
-      ): URIO[R2, GraphQLResponse[E2]] = f(self.execute(query, operationName, variables, skipValidation))
-      override def render: String      = self.render
-    }
+  final def |+|[R1 <: R](that: GraphQL[R1]): GraphQL[R1] = combine(that)
 
   /**
-   * Attaches a function that will analyze each query before execution, possibly modify or reject it.
-   * @param queryAnalyzer a function from `Field` to `ZIO[R, CalibanError, Field]`
-   * @return  a new GraphQL interpreter
+   * Renames the root queries, mutations and subscriptions objects.
+   * @param queriesName a new name for the root queries object
+   * @param mutationsName a new name for the root mutations object
+   * @param subscriptionsName a new name for the root subscriptions object
+   * @return a new GraphQL API
    */
-  def withQueryAnalyzer[R2 <: R](queryAnalyzer: QueryAnalyzer[R2]): GraphQL[R2, Q, M, S, E] =
-    new GraphQL[R2, Q, M, S, E] {
-      override def check(query: String): IO[CalibanError, Unit] = self.check(query)
-      override def execute[R1 <: R2](
-        query: String,
-        operationName: Option[String],
-        variables: Map[String, InputValue],
-        skipValidation: Boolean,
-        queryAnalyzers: List[QueryAnalyzer[R1]]
-      ): URIO[R1, GraphQLResponse[E]] =
-        self.execute(query, operationName, variables, skipValidation, queryAnalyzer :: queryAnalyzers)
-      override def render: String = self.render
-    }
+  final def rename(
+    queriesName: Option[String] = None,
+    mutationsName: Option[String] = None,
+    subscriptionsName: Option[String] = None
+  ): GraphQL[R] = new GraphQL[R] {
```

Comment:
This is neat, but I'm wondering if it could be potentially split off into something more general purpose traversing transformer. Of course it doesn't stop it from going in now and updating it later. But doing something like: https://www.apollographql.com/docs/apollo-server/features/schema-transforms/ might be compelling enough to hold off while fleshing this out.

Reply:
I agree with you. Currently it's not doable easily because the type definition and the resolver are 2 different objects (`__Type` and `Step`) and any change in one of them should also be applied to the other one (e.g. rename or remove fields). Both are trees but they are quite different: `__Type` is entirely known, while `Step` might contain effects that return other steps, etc. I think it requires deeper changes to make this possible.

This `rename` here is more a "quick fix" for people who are not happy with the root names that are used when you combine 2 APIs. It only affect documentation since queries don't include root type names.

I definitely want to keep thinking about a way to support that, so I'm gonna create an issue for it. I'd love to hear how you use the schema transform feature btw, here or on discord. Having ideas about concrete use cases would help.